### PR TITLE
Bug 1880941: kube-apiserver: log non-probe requests before ready

### PR DESCRIPTION
**Pull request by sttts**
Log and create NonReadyRequests event.

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go [hunk 1]
```diff
@@ -253,7 +263,25 @@ func handleRootHealthz(checks ...HealthChecker) http.HandlerFunc {
 		}
 
 		individualCheckOutput.WriteTo(w)
-		fmt.Fprint(w, "healthz check passed\n")
+		fmt.Fprintf(w, "%s check passed\n", name)
+
+		// signal first time this is healthy
+		select {
+		case <-hasBeenHealthy:
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:270]: what do we need this block? Is `sync.Once` not thread-safe ?

File: staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go [hunk 2]
```diff
@@ -100,3 +94,53 @@ func WithLateConnectionFilter(handler http.Handler) http.Handler {
...
+			if _, closed := <-hasBeenReadyCh; closed {
+				handler.ServeHTTP(w, r)
+				return
+			}
+		}
+
+		// ignore connections to local IP. Those clients better know what they are doing.
+		if pth := "/" + strings.TrimLeft(r.URL.Path, "/"); pth != "/readyz" && pth != "/healthz" && pth != "/livez" {
+			if isLocal(r) {
+				klog.V(2).Infof("Loopback request to %q (user agent %q). This client probably does not watch /readyz and might get inconsistent answers.", r.URL.Path, r.UserAgent())
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:113]: why not as `klog.Warningf` ?

File: staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:270]: let me just use sync.Once here. I think we don't have to optimize like I tried here.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:110]: copy & paste. But I agree.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:110]: Actually, no. `V(2).Info` is just right. We still have few clients that make use of this (disaster recovery for example). It's not a warning, just an info.

File: staging/src/k8s.io/apiserver/pkg/server/config_test.go [hunk 3]
```diff
@@ -275,6 +284,7 @@ func TestAuthenticationAuditAnnotationsDefaultChain(t *testing.T) {
 		RequestInfoResolver:   &request.RequestInfoFactory{},
 		RequestTimeout:        10 * time.Second,
 		LongRunningFunc:       func(_ *http.Request, _ *request.RequestInfo) bool { return false },
+		Serializer:            serializer.NewCodecFactory(scheme),
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/config_test.go:287]: not needed anymore?

File: staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go [hunk 4]
```diff
@@ -100,3 +91,45 @@ func WithLateConnectionFilter(handler http.Handler) http.Handler {
 		handler.ServeHTTP(w, r)
 	})
 }
+
+// WithNonReadyRequestLogging rejects the request until the process has been ready once.
+func WithNonReadyRequestLogging(handler http.Handler, hasBeenReadyCh <-chan struct{}, s runtime.NegotiatedSerializer) http.Handler {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:96]: s runtime.NegotiatedSerializer not used ?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:96]: not anymore

File: staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go [hunk 5]
```diff
@@ -100,3 +90,45 @@ func WithLateConnectionFilter(handler http.Handler) http.Handler {
...
 	})
 }
+
+// WithNonReadyRequestLogging rejects the request until the process has been ready once.
+func WithNonReadyRequestLogging(handler http.Handler, hasBeenReadyCh <-chan struct{}) http.Handler {
+	var nonReadyRequestReceived atomic.Bool
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if hasBeenReadyCh != nil {
+			if _, closed := <-hasBeenReadyCh; closed {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:100]: I think it will block when the `hasBeenReadyCh` wasn't closed

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:100]: it does. Fixing.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go:100]: fixed

**Comment by sttts** [comment]
Broken base images 😞

**Pull request merged**